plutus-core/plutus-core/test/CostModelSafety/Spec.hs

```diff
@@ -0,0 +1,209 @@
+import PlutusCore qualified as PLC
+import PlutusCore.Builtin
+import PlutusCore.Crypto.BLS12_381.G1 qualified as BLS12_381.G1
+import PlutusCore.Crypto.BLS12_381.G2 qualified as BLS12_381.G2
+import PlutusCore.Crypto.BLS12_381.Pairing qualified as BLS12_381.Pairing
+import PlutusCore.Data (Data (..))
+import PlutusCore.Evaluation.Machine.ExBudget (ExBudget (ExBudget))
+import PlutusCore.Evaluation.Machine.ExBudgetingDefaults (defaultBuiltinCostModel,
+                                                          defaultCekMachineCosts)
+import PlutusCore.Evaluation.Machine.ExBudgetStream (sumExBudgetStream)
+import PlutusCore.Evaluation.Machine.ExMemoryUsage (LiteralByteSize)
+import UntypedPlutusCore.Evaluation.Machine.Cek.CekMachineCosts (CekMachineCosts,
+                                                                 CekMachineCostsBase (..))
+
+import Data.ByteString qualified as BS
+import Data.Functor.Identity (Identity (..))
+import Data.Kind qualified as GHC (Type)
+import Data.List.Extra (enumerate)
+import Data.Text (Text)
+import Data.Word (Word8)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (assertBool, testCase)
+import Type.Reflection (TypeRep, eqTypeRep, pattern App, typeRep, (:~~:) (..))
+
+-- Machine costs
+checkBudget :: Identity ExBudget  -> IO ()
+checkBudget (Identity (ExBudget cpu mem)) = do
+  assertBool "exBudgetCPU  <= 0 in CEK machine costs" $ cpu > 0
+  assertBool "exBudgetMemory <= 0 in CEK machine costs" $ mem > 0
+
+-- Check that the machine costs are all strictly positive.  All of the fields are matched explicitly
+-- to make sure that we don't forget any new ones that get added.
+testMachineCostModel :: CekMachineCosts -> IO ()
+testMachineCostModel (
+  CekMachineCostsBase
+    cekStartupBudget
+    cekVarBudget
+    cekConstBudget
+    cekLamBudget
+    cekDelayBudget
+    cekForceBudget
+    cekApplyBudget
+    cekBuiltinBudget
+    cekConstrBudget
+    cekCaseBudget
+  ) = do
+    checkBudget cekStartupBudget
+    checkBudget cekVarBudget
+    checkBudget cekConstBudget
+    checkBudget cekLamBudget
+    checkBudget cekDelayBudget
+    checkBudget cekForceBudget
+    checkBudget cekApplyBudget
+    checkBudget cekBuiltinBudget
+    checkBudget cekConstrBudget
+    checkBudget cekCaseBudget
+
+-- Builtin costs
+
+{- Much of the code here is based on Evaluation.Spec and
+   PlutusCore.Generators.Hedgehog.Builtin.  The tests are complicated by the
+   fact that We can't just read the models from `builtinCostModels.json` and
+   check them since the actual models are defined by calls to `toBuiltinMeaning`
+   in PlutusCore.Default.Builtins and those are free to supply their own costing
+   functions and ignore the definitions in the JSON file.  The costing functions
+   end up in the `BuiltinMeanings` and we can't access them directly and feed
+   them suitable input sizes: instead we have to manufacture terms containing
+   constants of the desired sizes and feed them to the denotation of the
```

Comment:
I took a quick look, I think I could expose the raw costing functions in `BulitinMeaning`. Feel free to create a GitHub issue for me to do that if you want that functionality.

Reply:
It might be useful to have, but maybe in fact doing the tests this way is kind of helpful, since it means that the costing functions always get the arguments they'll get in real life: integers can't have size 0 for example, so we may not want to check some costing functions at 0.  I'm not sure about that.